tools/collapse_schemas/collapse-schemas.sh

```diff
@@ -0,0 +1,104 @@
+  exit 1
+fi
+
+execute_sql() {
+  local sql="$1"
+  local server="$2"
+
+  hab pkg exec core/postgresql psql -t -U hab -h 127.0.0.1 -p 5433 -c "$sql" "builder_$server"
+}
+
+process_shards() {
+  local upper_bound="$1"
+  local server="$2"
+  local do_it="$3"
+
+  declare -A count_map
+
+  mapfile -t shards < <(seq 0 "$upper_bound")
+
+  for shard in "${shards[@]}"
+  do
+    current_schema="shard_$shard"
+    # ordering by table_name descending gets us the "origins" table first, which is a pre-req for all the other tables due to foreign key constraints
+    tables=$(execute_sql "SELECT table_name FROM information_schema.tables WHERE table_schema='$current_schema' AND table_type='BASE TABLE' ORDER BY table_name DESC;" "$server")
+    echo "current schema = $current_schema"
+
+    for table in $tables
+    do
+      # migrations will run automatically for the public schema, so we don't need to transfer any data there.
+      if [ "$table" == "__diesel_schema_migrations" ]; then
+        continue
+      fi
+
+      # turns out when we pipe a value in to xargs w/o any flags, it trims the leading and trailing whitespace for us. handy!
+      count=$(execute_sql "SELECT COUNT(*) FROM $current_schema.$table;" "$server" | xargs)
+      echo "Count for $current_schema.$table = $count"
+
+      if [ -z "${count_map[$table]:-}" ]; then
+        count_map[$table]="$count"
+      else
+        new_total=$((count_map[$table] + count))
+        count_map[$table]="$new_total"
+      fi
+
+      if [ "$do_it" == "migrate" ]; then
+        execute_sql "INSERT INTO public.$table SELECT * FROM $current_schema.$table;" "$server"
+      fi
+    done
+  done
+
+  # double check that our counts are correct
+  if [ "$do_it" == "migrate" ]; then
+    errors=0
+    public_tables=$(execute_sql "SELECT table_name FROM information_schema.tables WHERE table_schema='public' AND table_type='BASE TABLE';" "$server")
+    for table in $public_tables
+    do
+      if [ "$table" == "__diesel_schema_migrations" ]; then
+        continue
+      fi
+
+      public_count=$(execute_sql "SELECT COUNT(*) FROM public.$table;" "$server" | xargs)
+
+      if [ -z "${count_map[$table]:-}" ]; then
+        echo "*** PANIC ***. There is no accumulated record total for the $table table. There should be. The total for the $table table in the public schema is $public_count."
+        errors=$((errors+1))
+      elif [ "${count_map[$table]}" -ne "$public_count" ]; then
+        echo "*** PANIC ***. The accumulated record total for the $table table across all shards is ${count_map[$table]}. That does not match the total record count in the public schema, which is $public_count."
+        errors=$((errors+1))
+      else
+        echo "The accumulated record count for the $table table matches the total record count in the public schema. All good."
+      fi
+    done
+
+    if [ "$errors" -eq 0 ]; then
+      echo "The entire shard migration completed without errors. Updating db flag."
```

Comment:
Minor - reword output to say to "schema migration" instead of "shard migration"?

Reply:
Part of the challenge here is the term "schema" is overloaded.  It can be used to refer to normal database schemas, e.g. the data definition of a table.  It can also be used to refer to a postgres specific entity that provides a logical separation between tables and functions, which is of course how our sharding works.  It feels important to me that we are careful to disambiguate between the two.  To a person that's not familiar with the intimate details of our sharding implementation, our regular database migrations are definitely "schema migrations".  I'd prefer to keep the language in this script specific to shards and not schemas, to eliminate any ambiguity.

Reply:
👍